internal/repository/validator.go

```diff
@@ -7,198 +7,339 @@ import (
-	ctx context.Context,
-	startBlock uint64,
-	endBlock uint64,
-) ([]Output, error) {
+// GetPreviousEpoch returns the epoch that ended one block before the start
+// of the current epoch
+func (pg *Database) GetPreviousEpoch(ctx context.Context, currentEpoch *Epoch) (*Epoch, error) {
+	if currentEpoch == nil {
+		return nil, NilEpoch
+	} else if currentEpoch.FirstBlock == 0 {
+		// if this is the first epoch, there is nothing to return
+		return nil, nil
+	}
+
 	query := `
 	SELECT
-		o.id,
-		o.index,
-		o.raw_data,
-		o.input_id,
-		i.status
+		id,
+    	application_address,
+        index,
+        first_block,
+		last_block,
+		claim_hash,
+		transaction_hash,
+		status
 	FROM
-		output o
-	INNER JOIN
-		input i
-	ON
-		o.input_id=i.id
+		epoch
 	WHERE
-		i.block_number BETWEEN @startBlock and @endBlock
+		application_address=@appAddress AND last_block=@lastBlock
 	ORDER BY
-		o.index asc`
+		index DESC
+	LIMIT 1
+	`
 
 	args := pgx.NamedArgs{
-		"startBlock": startBlock,
-		"endBlock":   endBlock,
+		"appAddress": currentEpoch.AppAddress,
+		"lastBlock":  currentEpoch.FirstBlock - 1,
 	}
 
-	rows, err := pg.db.Query(ctx, query, args)
-	if err != nil {
-		return nil, fmt.Errorf("Query failed: %v\n", err)
-	}
-
-	var id, input_id, index uint64
-	var rawData []byte
-	var status string
-	var results []Output
-
-	rowCount := 0
-
-	_, err = pgx.ForEachRow(rows, []any{&id, &index, &rawData, &input_id, &status},
-		func() error {
-			rowCount++
-			if status != string(InputStatusNone) {
-				output := Output{
-					Id:      id,
-					Index:   index,
-					RawData: rawData,
-					InputId: input_id,
-				}
-				results = append(results, output)
-			}
-			return nil
-		})
-	if err != nil {
-		return nil, fmt.Errorf("ForEachRow failed: %w\n", err)
-	}
+	var (
+		id, index, firstBlock, lastBlock uint64
+		appAddress                       Address
+		claimHash, transactionHash       *Hash
+		status                           EpochStatus
+	)
 
-	if len(results) == rowCount {
-		return results, nil
+	err := pg.db.QueryRow(ctx, query, args).Scan(
+		&id,
+		&appAddress,
+		&index,
+		&firstBlock,
+		&lastBlock,
+		&transactionHash,
+		&claimHash,
+		&status,
+	)
+	if err != nil {
+		return nil, fmt.Errorf("GetPreviousEpoch failed: %w\n", err)
 	}
 
-	return nil, nil
+	return &Epoch{
+		Id:              id,
+		Index:           index,
+		FirstBlock:      firstBlock,
+		LastBlock:       lastBlock,
+		TransactionHash: transactionHash,
+		ClaimHash:       claimHash,
+		Status:          status,
+		AppAddress:      appAddress,
+	}, nil
 }
 
+// SetEpochClaimAndInsertProofsTransaction performs a database transaction
+// containing two operations:
+//
+// 1. Updates an epoch, adding its claim and modifying its status.
+//
+// 2. Updates several outputs with their Keccak256 hash and proof.
 func (pg *Database) SetEpochClaimAndInsertProofsTransaction(
 	ctx context.Context,
-	epoch Epoch,
-	outputs []Output,
+	epoch *Epoch,
+	outputs []*Output,
```

Comment:
Do we need an array of pointers here? If there's a nil value in here, shouldn't we just remove it from the array?

Reply:
Slices copies share the underline array, right? So copying slices should not be expensive.

Using actual instances instead of pointers will prevent these possible caller errors